### PR TITLE
MFB-854 WA CSFP

### PR DESCRIPTION
**Pull request by elliottwaissbluth**
- [x] AI code draft written
 - [x] Coderabbit review
 - [x] Coderabbit comments addressed
 - [x] Claude review
 - [x] Claude comments addressed
 - [ ] Playwright validations

## Summary by CodeRabbit

* **New Features**
  * Introduced Washington Commodity Supplemental Food Program (CSFP) to the benefits screener with eligibility requirements including member age verification and household income assessment at 150% of federal poverty level thresholds.

### CODE REVIEW

**Comment by elliottwaissbluth** [comment]
# Code Review: MFB-854

**PR**: [#1484 - MFB-854 WA CSFP](https://github.com/MyFriendBen/benefits-api/pull/1484)
**Ticket**: [MFB-854 — WA: CSFP](https://linear.app/myfriendben/issue/MFB-854/wa-csfp)
**Author**: Elliott Waissbluth
**Type**: Backend (Django/Python) — New benefit program
**Reviewed**: 2026-05-06

---

## PR Overview

This PR implements the **Commodity Supplemental Food Program (CSFP)** for Washington State. CSFP is a federal USDA program that provides free monthly food commodity boxes to low-income seniors aged 60 and older. This is a new calculator added to the `wa` white label, giving eligible Washington residents a way to discover this food assistance program through the MyFriendBen screener.

### What Problem Does This Solve?

Seniors in Washington who qualify for CSFP aren't currently surfaced by the screener. This PR closes that gap by adding eligibility logic, program configuration, navigator contacts, and a comprehensive test suite for the program.

### How Does It Work?

#### Registers a new program in the WA calculator registry
- `programs/programs/wa/__init__.py` is updated to import `WaCsfp` and add it to `wa_calculators` under the key `"wa_csfp"`, making it discoverable by the dispatch layer.

#### Implements age- and income-based eligibility
- `programs/programs/wa/csfp/calculator.py` defines `WaCsfp`, a `ProgramCalculator` subclass with two eligibility checks:
  - **Member-level**: each household member must be age 60+
  - **Household-level**: annual gross income must be ≤ 150% of FPL for the household size
- The benefit value is `$50/month` per qualifying member, so a household with two seniors aged 60+ receives `$100/month`.

#### Seeds program data via the import config
- `programs/management/commands/import_program_config_data/data/wa_csfp_initial_config.json` provides the full program metadata (name, description, legal statuses, documents, and navigator contacts for four local CSFP agencies across WA).

#### Documents data gaps and eligibility criteria formally
- `programs/programs/wa/csfp/spec.md` records the full eligibility specification with source citations, noting two data gaps that can't be screened: geographic service area coverage and FDPIR dual-participation exclusion.

#### Provides unit tests and a validation fixture
- `programs/programs/wa/csfp/tests/test_wa_csfp.py` covers class attributes, member eligibility boundaries (age 59/60/68/None), household income boundaries, and value calculations for single and multi-senior households.
- `validations/management/commands/import_validations/data/wa_csfp.json` contains 3 validation scenarios (eligible senior, boundary eligible, age-ineligible) for the automated validation harness.

---

## Visual Architecture

### Calculator Hierarchy

```mermaid
classDiagram
    class ProgramCalculator {
        +Screen screen
        +Program program
        +dependencies: tuple
        +amount: int
        +member_amount: int
        +calc() Eligibility
        +eligible() Eligibility
        +member_eligible(e: MemberEligibility)
        +household_eligible(e: Eligibility)
        +value(e: Eligibility)
    }

    class WaCsfp {
        +min_age: int = 60
        +fpl_percent: float = 1.5
        +member_amount: int = 50
        +dependencies: list
        +member_eligible(e: MemberEligibility)
        +household_eligible(e: Eligibility)
    }

    class WaSsdi {
        +sga_non_blind: int
        +sga_blind: int
        +member_eligible(e)
        +household_eligible(e)
    }

    class WaWsosGrd {
        +amount: int = 25000
        +member_eligible(e)
        +household_eligible(e)
    }

    ProgramCalculator <|-- WaCsfp
    ProgramCalculator <|-- WaSsdi
    ProgramCalculator <|-- WaWsosGrd

    class wa_calculators {
        wa_csfp → WaCsfp
        wa_ssdi → WaSsdi
        wa_wsos_grd → WaWsosGrd
    }

    WaCsfp ..> wa_calculators : registered
```

### Eligibility Decision Flow

```mermaid
flowchart TD
    A[Screen submitted for WA white label] --> B{CSFP in wa_calculators?}
    B -->|Yes| C[WaCsfp.calc called]
    C --> D{Missing dependencies?}
    D -->|Yes| E[DependencyError raised]
    D -->|No| F[eligible called]

    F --> G[Iterate household members]
    G --> H{member.age >= 60?}
    H -->|Yes| I[MemberEligibility = True]
    H -->|No / None| J[MemberEligibility = False]

    I --> K{At least one eligible member?}
    J --> K
    K -->|No| L[Eligibility = False]
    K -->|Yes| M[household_eligible called]

    M --> N{gross_income <= int 1.5 × FPL?}
    N -->|Yes| O[Eligibility = True]
    N -->|No| P[Eligibility = False + income message]

    O --> Q[value called]
    Q --> R[$50 per eligible member]
    R --> S[Return Eligibility result]
    L --> S
    P --> S
```

### Data Seeding Flow

```mermaid
sequenceDiagram
    participant DevOps
    participant ImportCommand as manage.py import_program_config_data
    participant DB as PostgreSQL
    participant Screener as Screener API

    DevOps->>ImportCommand: Run with wa_csfp_initial_config.json
    ImportCommand->>DB: Upsert WhiteLabel(code=wa)
    ImportCommand->>DB: Upsert ProgramCategory(wa_food)
    ImportCommand->>DB: Create/Update Program(wa_csfp)
    ImportCommand->>DB: Create Documents (id_proof, wa_home, wa_proof_income)
    ImportCommand->>DB: Create Navigators (WSDA, OIC of WA, Partners Inland NW, Rural Resources CA)
    DB-->>ImportCommand: Saved
    Screener->>DB: Query programs for WA screen
    DB-->>Screener: Includes wa_csfp with full metadata
```

---

## Code Review

### 1. Income Limit Uses `int()` Truncation — Should Use `round()`

**`programs/programs/wa/csfp/calculator.py`, line 165**

The income limit is computed by multiplying `fpl_percent` (1.5) by the FPL dollar amount returned from `get_limit()`. When `get_limit()` returns an odd integer (which is common — FPL values are not always evenly divisible by 2), multiplying by 1.5 produces a `.5` fractional result. Wrapping in `int()` **truncates toward zero**, silently making the effective limit 50 cents lower than the actual 150% FPL threshold. This means a household whose income is exactly at the legal limit could be told they're over it.

`WaWsosGrd` has an explicit comment calling this out: _"compare against the limit at full precision so a household whose annual income is fractionally above the … cap is correctly screened out instead of being floored down onto the boundary."_ The pattern there avoids any float-to-int conversion for the limit itself (the MFI table values are already whole dollars).

```diff
  def household_eligible(self, e: Eligibility):
      gross_income = self.screen.calc_gross_income("yearly", ["all"])
-     income_limit = int(self.fpl_percent * self.program.year.get_limit(self.screen.household_size))
+     income_limit = round(self.fpl_percent * self.program.year.get_limit(self.screen.household_size))
      e.condition(gross_income <= income_limit, messages.income(gross_income, income_limit))
```

`round()` ties to even at 0.5 (banker's rounding), which is the most defensible behavior for a benefits boundary. Alternatively, `math.ceil()` could be used to always round the limit up (more favorable to applicants).

---

### 2. No `has_benefit("csfp")` Guard — Current Recipients Are Shown as Eligible

**`programs/programs/wa/csfp/calculator.py`, line 163**

The spec (Scenario 10) calls out that a current CSFP participant should be shown as **not eligible** because 7 CFR 247.9(c) prohibits re-enrollment while already participating. The `WaSsdi` calculator demonstrates the pattern:

```python
# WaSsdi.household_eligible (lines 79–82)
e.condition(
    not self.screen.has_benefit("ssdi"),
    messages.must_not_have_benefit("SSDI"),
)
```

`WaCsfp.household_eligible` has no equivalent check. A screen where the user selects CSFP as a current benefit will still calculate as eligible (if age and income criteria are met), which conflicts with both the spec and federal regulations.

```diff
  def household_eligible(self, e: Eligibility):
      gross_income = self.screen.calc_gross_income("yearly", ["all"])
      income_limit = round(self.fpl_percent * self.program.year.get_limit(self.screen.household_size))
+     e.condition(
+         not self.screen.has_benefit("csfp"),
+         messages.must_not_have_benefit("CSFP"),
+     )
      e.condition(gross_income <= income_limit, messages.income(gross_income, income_limit))
```

After adding this check, a corresponding test should be added:

```diff
  # In TestWaCsfpHouseholdEligibility (or a new TestWaCsfpCurrentBenefits class)
+ def test_already_receiving_csfp_ineligible(self):
+     calc = make_calculator(yearly_income=10_000)
+     calc.screen.has_benefit = Mock(return_value=True)
+     e = Eligibility()
+     me = MemberEligibility(make_member())
+     me.eligible = True
+     e.add_member_eligibility(me)
+     calc.household_eligible(e)
+     self.assertFalse(e.eligible)
```

---

### 3. Minor — Test `make_calculator` Does Not Stub `household_size` Explicitly (Low Risk)

**`programs/programs/wa/csfp/tests/test_wa_csfp.py`, lines 461–472**

`self.screen.household_size` is passed as an argument to `mock_program.year.get_limit()`. Because `get_limit` is a `Mock(return_value=fpl_limit)` that ignores its argument, this is harmless today. But if the calculator is ever refactored to use `household_size` directly (e.g., to look up an additional-person multiplier), the tests will silently pass a `MagicMock` object instead of an integer.

This is low priority, but adding an explicit stub would make tests more intention-revealing:

```diff
  def make_calculator(yearly_income=10_000, fpl_limit=15_000, members=None):
      mock_screen = Mock()
      mock_screen.calc_gross_income = Mock(return_value=yearly_income)
      mock_screen.household_members.all = Mock(return_value=members or [make_member()])
+     mock_screen.household_size = 1
```

---

### 4. Positive — Data Gap Documentation is Excellent

**`programs/programs/wa/csfp/calculator.py`, lines 133–146**

The class docstring clearly names the two criteria that can't be evaluated (geographic service area, FDPIR exclusion), explains why each gap exists, cites the federal regulation, and assigns an impact level. This is exactly the right level of documentation — a future maintainer can instantly tell what the screener does and doesn't check, and why.

No change needed.

---

### 5. Positive — `__init__.py` Trailing Comma

**`programs/programs/wa/__init__.py`, line 116**

The trailing comma added to the `WaWsosGrd` entry is a good cleanup that makes future additions cleaner. No issue.

---

## Security Analysis

- **Input Validation**: PASS — All inputs (`age`, income) are handled through `Screen.calc_gross_income` and `HouseholdMember.age`, both of which are sanitized at the model/serializer layer before reaching this calculator.
- **SQL Injection**: PASS — No raw SQL. All DB access goes through the ORM via established patterns (`get_limit`, `calc_gross_income`).
- **XSS Prevention**: PASS — Calculator returns structured Python objects (no HTML output).
- **Authentication/Authorization**: PASS — White label routing (`wa` only) is enforced by the `wa_calculators` registry; no new endpoints introduced.
- **Sensitive Data**: PASS — No PII in code; navigator email/phone in the config JSON is public contact info for government agencies.
- **CSRF Protection**: PASS — No state-changing HTTP endpoints introduced.
- **Dependencies**: PASS — No new third-party dependencies.

**Critical Issues**: None.

**Recommendations**: None beyond the `has_benefit` guard above (which is a correctness issue, not a security vulnerability).

---

## Impact Analysis

### What Changed

| File | Change |
|------|--------|
| `programs/programs/wa/__init__.py` | Added `WaCsfp` import and registry entry |
| `programs/programs/wa/csfp/calculator.py` | New calculator (40 lines) |
| `programs/programs/wa/csfp/tests/test_wa_csfp.py` | New test module (129 lines) |
| `programs/programs/wa/csfp/spec.md` | New spec (265 lines) |
| `validations/.../wa_csfp.json` | New validation fixture (130 lines) |
| `programs/.../wa_csfp_initial_config.json` | New program seed data (95 lines) |

### Potential Impact

**Direct Dependencies:**
- Any screen submitted under the `wa` white label will now evaluate `wa_csfp` eligibility as part of its program list — this is the intended effect, not a risk.
- The `wa_calculators` dict is now 3 entries; no other code iterates this dict with assumptions about count.

**Indirect Dependencies:**
- `import_program_config_data` management command: the new JSON will be imported on the next data seed run. No existing WA data is modified.
- The validation harness will pick up `wa_csfp.json` on its next run.

**Breaking Changes**: None. All changes are purely additive.

### Recommended Additional Testing

- Manual: run a WA screen with a 60+ member and income below 150% FPL → confirm CSFP appears
- Manual: run a WA screen where CSFP is selected as a current benefit → confirm CSFP does *not* appear (blocked pending fix #2 above)
- Manual: run a WA screen with all members under 60 → confirm CSFP does not appear
- Regression: verify that existing WA programs (`wa_ssdi`, `wa_wsos_grd`) are unaffected

---

## Business Logic Correctness

**Spec compared**: `programs/programs/wa/csfp/spec.md`

**Correctly implemented:**
- Criterion 1 — Age ≥ 60: `e.condition(member.age is not None and member.age >= self.min_age)` ✅
- Criterion 2 — Income ≤ 150% FPL: `e.condition(gross_income <= income_limit, ...)` ✅ *(with int-truncation caveat noted above)*
- Criterion 3 — WA residency: enforced implicitly by `wa` white label routing ✅
- Benefit value — ~$50/month per eligible member: `member_amount = 50` ✅
- Data gaps documented — service area and FDPIR: class docstring ✅

**Discrepancies found:**

| Criterion | Spec Says | Code Does |
|-----------|-----------|-----------|
| Criterion 5 — FDPIR exclusion | Must not participate in FDPIR; acknowledged as data gap | Not checked; documented in docstring as data gap ✅ (acceptable) |
| Scenario 10 — Already receiving CSFP | Should be ineligible (7 CFR 247.9(c)) | No `has_benefit("csfp")` check; will show as eligible ⚠️ |
| Income limit precision | Legal threshold is 150% FPL exactly | `int()` truncation can understate limit by $0.50 ⚠️ |

**Spec rules not covered:**
- The `has_benefit("csfp")` exclusion (Scenario 10) is the only evaluable rule missing from the implementation.

---

## Migration Safety

N/A — no database migrations in this PR.

---

## Checklist

- [x] Code follows project conventions (Django/Python patterns)
- [x] Tests are comprehensive and cover boundary conditions
- [x] No security vulnerabilities introduced
- [x] No breaking changes
- [ ] Business logic fully matches program spec — `has_benefit("csfp")` check missing
- [x] Database queries are optimized (no N+1; uses existing screener methods)
- [x] Data gaps documented with citations
- [x] Proper white label scoping via `wa_calculators` registry
- [ ] Income limit precision — `int()` should be `round()`
- [x] Validation fixture provided (3 scenarios)

---

## Summary

**Recommendation**: REQUEST CHANGES (minor — two targeted fixes needed)

**Key Takeaways:**
1. The calculator is clean, well-documented, and correctly structured — the eligibility logic follows the established pattern and data gaps are clearly flagged with regulatory citations.
2. **Missing `has_benefit("csfp")` check**: spec Scenario 10 and 7 CFR 247.9(c) require that current CSFP participants are not shown the program as available; adding `e.condition(not self.screen.has_benefit("csfp"), messages.must_not_have_benefit("CSFP"))` in `household_eligible` is a five-line fix.
3. **`int()` truncation on income limit**: using `round()` instead of `int()` aligns with the precision approach used elsewhere in the codebase (see `WaWsosGrd` comment) and prevents edge cases at the 150% FPL boundary.

**Next Steps:**
1. Add `has_benefit("csfp")` guard in `household_eligible` + corresponding test
2. Replace `int()` with `round()` for the income limit calculation
3. Run the validation harness (`wa_csfp.json`) and confirm all 3 scenarios pass
4. Playwright smoke test: confirm CSFP surfaces correctly on a WA screen

---

*Review generated by Claude Code — claude-sonnet-4-6*
*Linear Ticket: [MFB-854](https://linear.app/myfriendben/issue/MFB-854/wa-csfp)*
*GitHub PR: [#1484](https://github.com/MyFriendBen/benefits-api/pull/1484)*

File: programs/programs/wa/csfp/spec.md [hunk 1]
```diff
@@ -0,0 +1,265 @@
...
+**Expected**: Not eligible, value: `$0`
+
+**Steps**:
+
+* **Location**: Enter ZIP code `98101`, Select county `King`
+* **Household**: Number of people: `1`
+* **Person 1**: Birth month/year: `March 1961` (age 65), This person is the head of household, Has income: Yes, Social Security Retirement income: `$1,999` per month (approximately $23,988/year, just above the 150% FPL limit of ~$23,475/year for a 1-person household using 2025 HHS guidelines), No other income sources
+* **Current Benefits**: Do not select any current benefits
+
+**Why this matters**: This test ensures the screener correctly rejects applicants whose income is just barely above the 150% FPL threshold (the federal limit effective January 2026), validating that the income boundary is enforced precisely rather than allowing a margin of error. N
```
[the remark this answers is not in the document]
Comment (elliottwaissbluth) [reply programs/programs/wa/csfp/spec.md:129]: Honestly I feel like this one is fine.
